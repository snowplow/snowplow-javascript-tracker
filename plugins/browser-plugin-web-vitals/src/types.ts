import type { ReportCallback, Metric, WebVitalsGlobal, onCLS, onLCP, onFID, onFCP, onINP, onTTFB } from 'web-vitals';

export interface WebVitals extends WebVitalsGlobal {
  onCLS: typeof onCLS;
  onFID: typeof onFID;
  onLCP: typeof onLCP;
  onFCP: typeof onFCP;
  onINP: typeof onINP;
  onTTFB: typeof onTTFB;
}

export { Metric, ReportCallback };
