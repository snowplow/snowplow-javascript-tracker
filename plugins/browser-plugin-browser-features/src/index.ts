/*
 * Copyright (c) 2022 Snowplow Analytics Ltd, 2010 Anthon Pang
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice, this
 *    list of conditions and the following disclaimer.
 *
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 *    this list of conditions and the following disclaimer in the documentation
 *    and/or other materials provided with the distribution.
 *
 * 3. Neither the name of the copyright holder nor the names of its
 *    contributors may be used to endorse or promote products derived from
 *    this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

import { isFunction, BrowserPlugin, BrowserTracker } from '@snowplow/browser-tracker-core';

declare global {
  interface MimeTypeArray {
    [index: string]: MimeType;
  }
}

const pluginMap: Record<string, string> = {
  // document types
  pdf: 'application/pdf',

  // media players
  qt: 'video/quicktime',
  realp: 'audio/x-pn-realaudio-plugin',
  wma: 'application/x-mplayer2',

  // interactive multimedia
  dir: 'application/x-director',
  fla: 'application/x-shockwave-flash',

  // RIA
  java: 'application/x-java-vm',
  gears: 'application/x-googlegears',
  ag: 'application/x-silverlight',
};

/**
 * Adds the available MIME Types to each event
 */
export function BrowserFeaturesPlugin(): BrowserPlugin {
  return {
    activateBrowserPlugin: (tracker: BrowserTracker) => {
      const navigatorAlias = navigator;
      // General plugin detection
      if (navigatorAlias.mimeTypes && navigatorAlias.mimeTypes.length) {
        for (const i in pluginMap) {
          if (Object.prototype.hasOwnProperty.call(pluginMap, i)) {
            let mimeType = navigatorAlias.mimeTypes[pluginMap[i]];
            if (mimeType) {
              tracker.core.addPayloadPair('f_' + i, mimeType.enabledPlugin ? '1' : '0');
            }
          }
        }
      }

      // Firefox
      if (isFunction((window as any).GearsFactory)) {
        tracker.core.addPayloadPair('f_gears', '1');
      }
    },
  };
}
