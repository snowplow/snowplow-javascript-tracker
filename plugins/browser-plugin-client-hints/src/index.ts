import { BrowserPlugin } from '@snowplow/browser-tracker-core';
import { HttpClientHints } from './contexts';

interface NavigatorUABrandVersion {
  brand: string; // "Google Chrome"
  version: string; // "84"
}

interface UADataValues {
  platform: string; // "PhoneOS"
  platformVersion: string; // "10A"
  architecture: string; // "arm"
  model: string; // "X644GTM"
  uaFullVersion: string; // "73.32.AGX.5"
}

declare global {
  interface Navigator {
    userAgentData: {
      mobile: boolean;
      brands: Array<NavigatorUABrandVersion>;
      getHighEntropyValues: (hints: Array<string>) => Promise<UADataValues>;
    };
  }
}

let uaClientHints: HttpClientHints;

/**
 * This function makes sure that the expected array is returned as an array instead of an object.
 * It handles a problem that in some cases the `navigator.userAgentData.brands` was returned as an object instead of array.
 */
function forceArray<T>(array: T[] | Record<string, T>): T[] {
  if (Array.isArray(array)) return array;

  if (Object.prototype.toString.call(array) === '[object Object]') {
    return Object.keys(array).map((e) => {
      return array[e];
    });
  }

  return [];
}

/**
 * Returns the client-hints brands, ensuring no additional properties.
 */
function getBrands(brands: Array<NavigatorUABrandVersion>): Array<NavigatorUABrandVersion> {
  return brands.map((b) => {
    const { brand, version } = b;
    return { brand, version };
  });
}

/**
 * Validates whether userAgentData is compliant to the client-hints interface.
 * https://wicg.github.io/ua-client-hints/#interface
 */
function validClientHints(hints: HttpClientHints): boolean {
  if (!hints || typeof hints.isMobile !== 'boolean' || !Array.isArray(hints.brands)) {
    return false;
  }

  if (
    hints.brands.length === 0 ||
    hints.brands.some((brand) => typeof brand.brand !== 'string' || typeof brand.version !== 'string')
  ) {
    return false;
  }

  return true;
}

/**
 * Attaches Client Hint information where available
 * @param includeHighEntropy - Should high entropy values be included
 */
export function ClientHintsPlugin(includeHighEntropy?: boolean): BrowserPlugin {
  const populateClientHints = () => {
    const navigatorAlias = navigator;
    const uaData = navigatorAlias.userAgentData;

    if (uaData) {
      let candidateHints: HttpClientHints;
      candidateHints = {
        isMobile: uaData.mobile,
        brands: getBrands(forceArray(uaData.brands)),
      };
      if (includeHighEntropy && uaData.getHighEntropyValues) {
        uaData
          .getHighEntropyValues(['platform', 'platformVersion', 'architecture', 'model', 'uaFullVersion'])
          .then((res) => {
            candidateHints.architecture = res.architecture;
            candidateHints.model = res.model;
            candidateHints.platform = res.platform;
            candidateHints.uaFullVersion = res.uaFullVersion;
            candidateHints.platformVersion = res.platformVersion;
          });
      }

      if (validClientHints(candidateHints)) {
        uaClientHints = candidateHints;
      }
    }
  };

  return {
    activateBrowserPlugin: () => {
      if (!uaClientHints) {
        populateClientHints();
      }
    },
    contexts: () => {
      if (uaClientHints) {
        return [
          {
            schema: 'iglu:org.ietf/http_client_hints/jsonschema/1-0-0',
            data: uaClientHints,
          },
        ];
      }
      return [];
    },
  };
}
