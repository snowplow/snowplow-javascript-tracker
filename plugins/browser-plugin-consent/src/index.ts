/*
 * Copyright (c) 2022 Snowplow Analytics Ltd, 2010 Anthon Pang
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice, this
 *    list of conditions and the following disclaimer.
 *
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 *    this list of conditions and the following disclaimer in the documentation
 *    and/or other materials provided with the distribution.
 *
 * 3. Neither the name of the copyright holder nor the names of its
 *    contributors may be used to endorse or promote products derived from
 *    this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

import { BrowserPlugin, BrowserTracker, dispatchToTrackersInCollection } from '@snowplow/browser-tracker-core';
import {
  buildConsentGranted,
  buildConsentWithdrawn,
  CommonEventProperties,
  ConsentGrantedEvent,
  ConsentWithdrawnEvent,
  Logger,
} from '@snowplow/tracker-core';
import { Gdpr } from './contexts';

export { ConsentGrantedEvent, ConsentWithdrawnEvent };

export enum gdprBasis {
  consent = 'consent',
  contract = 'contract',
  legalObligation = 'legal_obligation',
  vitalInterests = 'vital_interests',
  publicTask = 'public_task',
  legitimateInterests = 'legitimate_interests',
}

export type GdprBasis = keyof typeof gdprBasis;

/** The Configuration for the GDPR Context */
export interface GdprContextConfiguration {
  /** The basis for why the document will be processed */
  basisForProcessing: GdprBasis;
  /** An identifier for the document */
  documentId?: string;
  /** The version of the document */
  documentVersion?: string;
  /** A descrtiption of the document */
  documentDescription?: string;
}

const _trackers: Record<string, BrowserTracker> = {};
const _context: Record<string, Gdpr> = {};
let LOG: Logger;

/**
 * The Consent Plugin
 *
 * Adds Consent Granted and Withdrawn events
 * and the ability to add the GDPR context to events
 */
export function ConsentPlugin(): BrowserPlugin {
  let trackerId: string;

  return {
    activateBrowserPlugin: (tracker) => {
      trackerId = tracker.id;
      _trackers[tracker.id] = tracker;
    },
    contexts: () => {
      if (_context[trackerId]) {
        return [
          {
            schema: 'iglu:com.snowplowanalytics.snowplow/gdpr/jsonschema/1-0-0',
            data: _context[trackerId],
          },
        ];
      }

      return [];
    },
    logger: (logger) => {
      LOG = logger;
    },
  };
}

/**
 * Enable the GDPR context for each event
 * @param configuration - the configuration for the GDPR context
 * @param trackers - The tracker identifiers which should have the GDPR context enabled
 */
export function enableGdprContext(
  configuration: GdprContextConfiguration,
  trackers: Array<string> = Object.keys(_trackers)
) {
  const { basisForProcessing, documentId, documentVersion, documentDescription } = configuration;
  let basis = gdprBasis[basisForProcessing];

  if (!basis) {
    LOG.warn(
      'enableGdprContext: basisForProcessing must be one of: consent, contract, legalObligation, vitalInterests, publicTask, legitimateInterests'
    );
    return;
  } else {
    trackers.forEach((t) => {
      if (_trackers[t]) {
        _context[t] = {
          basisForProcessing: basis,
          documentId: documentId ?? null,
          documentVersion: documentVersion ?? null,
          documentDescription: documentDescription ?? null,
        };
      }
    });
  }
}

/**
 * Track a consent granted action
 *
 * @param event - The event information
 * @param trackers - The tracker identifiers which the event will be sent to
 */
export function trackConsentGranted(
  event: ConsentGrantedEvent & CommonEventProperties,
  trackers: Array<string> = Object.keys(_trackers)
) {
  dispatchToTrackersInCollection(trackers, _trackers, (t) => {
    const builtEvent = buildConsentGranted(event);
    t.core.track(
      builtEvent.event,
      event.context ? event.context.concat(builtEvent.context) : builtEvent.context,
      event.timestamp
    );
  });
}

/**
 * Track a consent withdrawn action
 *
 * @param event - The event information
 * @param trackers - The tracker identifiers which the event will be sent to
 */
export function trackConsentWithdrawn(
  event: ConsentWithdrawnEvent & CommonEventProperties,
  trackers: Array<string> = Object.keys(_trackers)
) {
  dispatchToTrackersInCollection(trackers, _trackers, (t) => {
    const builtEvent = buildConsentWithdrawn(event);
    t.core.track(
      builtEvent.event,
      event.context ? event.context.concat(builtEvent.context) : builtEvent.context,
      event.timestamp
    );
  });
}
