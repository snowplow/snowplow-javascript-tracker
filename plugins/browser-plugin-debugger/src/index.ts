/*
 * Copyright (c) 2022 Snowplow Analytics Ltd, 2010 Anthon Pang
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice, this
 *    list of conditions and the following disclaimer.
 *
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 *    this list of conditions and the following disclaimer in the documentation
 *    and/or other materials provided with the distribution.
 *
 * 3. Neither the name of the copyright holder nor the names of its
 *    contributors may be used to endorse or promote products derived from
 *    this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

import { BrowserPlugin, BrowserTracker } from '@snowplow/browser-tracker-core';
import {
  EventJson,
  EventJsonWithKeys,
  JsonProcessor,
  Logger,
  LOG_LEVEL,
  Payload,
  PayloadBuilder,
  payloadJsonProcessor,
  version,
  getSchemaParts,
  SelfDescribingJson,
} from '@snowplow/tracker-core';
import randomcolor from 'randomcolor';

/**
 * Adds advertisement tracking functions
 */
export function DebuggerPlugin(logLevel: LOG_LEVEL = LOG_LEVEL.debug): BrowserPlugin {
  let LOG: Logger;
  let tracker: BrowserTracker;
  let eventColour: string;

  const colours = {
    event: () => `color: White; background: ${eventColour}; font-weight: bold; padding: 1px 4px; border-radius: 2px;`,
    snowplowPurple: 'color: White; background: #6638B8; font-weight: bold; padding: 1px 4px; border-radius: 2px;',
    activated: 'color: White; background: #9E62DD; font-weight: bold; padding: 1px 4px; border-radius: 2px;',
    payload: 'color: White; background: #3748B8; font-weight: bold; padding: 1px 4px; border-radius: 2px;',
    json: 'color: White; background: #388AB8; font-weight: bold; padding: 1px 4px; border-radius: 2px;',
    schema: 'color: White; background: #268047; font-weight: bold; padding: 1px 4px; border-radius: 2px;',
    schemaVersion: 'color: White; background: #80265F; font-weight: bold; padding: 1px 4px; border-radius: 2px;',
  };

  const debug = (style: string, message: string, extra?: [string, ...unknown[]]) => {
    const [extraMessage, ...rest] = extra ?? [''];
    LOG.debug(
      `v${version} %c${tracker.namespace}%c %c%s` + extraMessage,
      colours.snowplowPurple,
      '',
      style,
      message,
      ...rest
    );
  };

  function jsonInterceptor(encodeBase64: boolean): JsonProcessor {
    const log = (json: EventJsonWithKeys, data: SelfDescribingJson) => {
      const schemaParts = getSchemaParts(data['schema']);
      debug(colours.event(), 'Event', [
        '%c%s%c%s%c%s\n%o',
        colours.json,
        `${getJsonType(json)}: ${schemaParts ? schemaParts[1] : 'Unknown Schema'}`,
        colours.schemaVersion,
        schemaParts ? `Version: ${schemaParts[2]}-${schemaParts[3]}-${schemaParts[4]}` : 'Unknown Schema Version',
        colours.schema,
        schemaParts ? `Vendor: ${schemaParts[0]}` : 'Unknown Vendor',
        data['data'],
      ]);
    };

    return (payloadBuilder: PayloadBuilder, jsonForProcessing: EventJson) => {
      if (jsonForProcessing.length) {
        for (const json of jsonForProcessing) {
          const data = json.json['data'] as SelfDescribingJson;
          if (Array.isArray(data)) {
            data.forEach((d) => {
              log(json, d);
            });
          } else {
            log(json, data);
          }
        }
      }
      return payloadJsonProcessor(encodeBase64)(payloadBuilder, jsonForProcessing);
    };
  }

  return {
    logger: (logger) => {
      LOG = logger;
      logger.setLogLevel(logLevel);
    },
    activateBrowserPlugin: (t) => {
      tracker = t;
      debug(colours.activated, 'Tracker Activated');
    },
    beforeTrack: (payloadBuilder: PayloadBuilder) => {
      eventColour = randomcolor({ luminosity: 'dark' });
      payloadBuilder.withJsonProcessor(jsonInterceptor(tracker.core.getBase64Encoding()));
      debug(colours.event(), 'Event', ['%c%s', colours.snowplowPurple, getEventType(payloadBuilder)]);
      payloadBuilder.build();
    },
    afterTrack: (payload: Payload) => {
      debug(colours.event(), 'Event', ['%c%s\n%o', colours.payload, 'Payload', payload]);
    },
  };
}

function getJsonType(json: EventJsonWithKeys) {
  switch (json.keyIfEncoded) {
    case 'cx':
      return 'Context';
    case 'ue_px':
      return 'Self Describing';
    default:
      return `${json.keyIfEncoded}, ${json.keyIfNotEncoded}`;
  }
}

function getEventType(payloadBuilder: PayloadBuilder) {
  const payload = payloadBuilder.getPayload();
  switch (payload['e']) {
    case 'pv':
      return 'Page View';
    case 'pp':
      return 'Page Ping';
    case 'tr':
      return 'Ecommerce Transaction';
    case 'ti':
      return 'Ecommerce Transaction Item';
    case 'se':
      return 'Structured Event';
    case 'ue':
      return 'Self Describing';
    default:
      return typeof payload['e'] === 'string' ? payload['e'] : 'Invalid';
  }
}
