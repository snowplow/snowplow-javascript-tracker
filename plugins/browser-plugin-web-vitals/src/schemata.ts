export const WEB_VITALS_SCHEMA = 'iglu:com.snowplowanalytics.snowplow/web_vitals/jsonschema/1-0-0';
