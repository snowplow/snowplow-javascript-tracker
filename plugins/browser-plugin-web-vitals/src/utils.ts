import { LOG } from '@snowplow/tracker-core';
import { ReportCallback, WebVitals } from './types';

/**
 * Attach page listeners to collect the Web Vitals values
 * @param {() => void} callback
 */
export function attachWebVitalsPageListeners(callback: () => void) {
  // Safari does not fire "visibilitychange" on the tab close
  // So we have 2 options: lose Safari data, or lose LCP/CLS that depends on "visibilitychange" logic.
  // Current solution: if LCP/CLS supported, use `onHidden` otherwise, use `pagehide` to fire the callback in the end.
  //
  // More details: https://github.com/treosh/web-vitals-reporter/issues/3
  const supportedEntryTypes = (PerformanceObserver && PerformanceObserver.supportedEntryTypes) || [];
  const isLatestVisibilityChangeSupported = supportedEntryTypes.indexOf('layout-shift') !== -1;

  if (isLatestVisibilityChangeSupported) {
    const onVisibilityChange = () => {
      if (document.visibilityState === 'hidden') {
        callback();
        window.removeEventListener('visibilitychange', onVisibilityChange, true);
      }
    };
    window.addEventListener('visibilitychange', onVisibilityChange, true);
  } else {
    window.addEventListener('pagehide', callback, { capture: true, once: true });
  }
}

/**
 *
 * @param {string} webVitalsSource Web Vitals script source.
 * @returns {string} The script element of the Web Vitals script. Used for attaching listeners on it.
 */
export function createWebVitalsScript(webVitalsSource: string) {
  const webVitalsScript = document.createElement('script');
  webVitalsScript.setAttribute('src', webVitalsSource);
  webVitalsScript.setAttribute('async', '1');
  webVitalsScript.addEventListener('error', () => {
    LOG.error(`Failed to load ${webVitalsSource}`);
  });

  document.head.appendChild(webVitalsScript);
  return webVitalsScript;
}

/**
 *
 * Adds the Web Vitals measurements on the object used by the trackers to store metric properties.
 * @param {Record<string, unknown>} webVitalsObject
 * @return {void}
 */
export function webVitalsListener(webVitalsObject: Record<string, unknown>) {
  function addWebVitalsMeasurement(metricSchemaName: string): ReportCallback {
    return (arg) => {
      webVitalsObject[metricSchemaName] = arg.value;
      webVitalsObject.navigationType = arg.navigationType;
    };
  }
  if (!window.webVitals) {
    LOG.warn('The window.webVitals API is currently unavailable. web_vitals events will not be collected.');
    return;
  }

  const webVitals = window.webVitals as WebVitals;
  webVitals.onCLS(addWebVitalsMeasurement('cls'));
  webVitals.onFID(addWebVitalsMeasurement('fid'));
  webVitals.onLCP(addWebVitalsMeasurement('lcp'));
  webVitals.onFCP(addWebVitalsMeasurement('fcp'));
  webVitals.onINP(addWebVitalsMeasurement('inp'));
  webVitals.onTTFB(addWebVitalsMeasurement('ttfb'));
}
