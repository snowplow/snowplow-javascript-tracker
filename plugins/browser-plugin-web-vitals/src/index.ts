import { BrowserPlugin, BrowserTracker, dispatchToTrackersInCollection } from '@snowplow/browser-tracker-core';
import { buildSelfDescribingEvent } from '@snowplow/tracker-core';
import { WEB_VITALS_SCHEMA } from './schemata';
import { attachWebVitalsPageListeners, createWebVitalsScript, webVitalsListener } from './utils';

const _trackers: Record<string, BrowserTracker> = {};
const WEB_VITALS_SOURCE = 'https://unpkg.com/web-vitals@3/dist/web-vitals.iife.js';
let listenersAttached = false;

interface WebVitalsPluginOptions {
  loadWebVitalsScript?: boolean;
  webVitalsSource?: string;
}

const defaultPluginOptions = {
  loadWebVitalsScript: true,
  webVitalsSource: WEB_VITALS_SOURCE,
};

/**
 * Adds Web Vitals measurement events
 *
 * @param pluginOptions.loadWebVitalsScript - Should the plugin immediately load the Core Web Vitals measurement script from UNPKG CDN.
 * @param pluginOptions.webVitalsSource - The URL endpoint the Web Vitals script should be loaded from. Defaults to the UNPKG CDN.
 * @remarks
 */
export function WebVitalsPlugin(pluginOptions: WebVitalsPluginOptions = defaultPluginOptions): BrowserPlugin {
  const webVitalsObject: Record<string, unknown> = {};
  const options = { ...defaultPluginOptions, ...pluginOptions };
  let trackerId: string;
  let webVitalsScript: HTMLScriptElement | undefined;
  return {
    activateBrowserPlugin: (tracker) => {
      trackerId = tracker.id;
      _trackers[trackerId] = tracker;

      function sendWebVitals() {
        if (!Object.keys(webVitalsObject).length) {
          return;
        }
        dispatchToTrackersInCollection(Object.keys(_trackers), _trackers, (t) => {
          t.core.track(
            buildSelfDescribingEvent({
              event: {
                schema: WEB_VITALS_SCHEMA,
                data: webVitalsObject,
              },
            })
          );
        });
      }

      if (options.loadWebVitalsScript) {
        webVitalsScript = createWebVitalsScript(options.webVitalsSource);
      }

      /*
       * Attach page listeners only once per page.
       * Prevent multiple trackers from attaching listeners multiple times.
       */
      if (!listenersAttached) {
        if (webVitalsScript) {
          webVitalsScript.addEventListener('load', () => webVitalsListener(webVitalsObject));
        } else {
          webVitalsListener(webVitalsObject);
        }
        attachWebVitalsPageListeners(sendWebVitals);
        listenersAttached = true;
      }

      return;
    },
  };
}
