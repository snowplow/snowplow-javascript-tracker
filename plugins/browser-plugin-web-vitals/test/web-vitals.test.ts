import { JSDOM } from 'jsdom';
import { trackerCore } from '@snowplow/tracker-core';
import { WebVitalsPlugin } from '../src';
import { BrowserTracker } from '@snowplow/browser-tracker-core';

declare var jsdom: JSDOM;

// @ts-expect-error
jsdom.window.webVitals = {};

describe('Web Vitals plugin', () => {
  function webVitalsCallback(callback: any) {
    callback({
      value: 0.01,
      navigationType: 'navigation',
    });
    return;
  }
  it('Returns values for Web Vitals properties', (done) => {
    Object.defineProperty(jsdom.window, 'PerformanceObserver', { value: jest.fn() });
    Object.defineProperty(jsdom.window.webVitals, 'onCLS', {
      value: webVitalsCallback,
    });
    Object.defineProperty(jsdom.window.webVitals, 'onLCP', {
      value: webVitalsCallback,
    });
    Object.defineProperty(jsdom.window.webVitals, 'onFCP', {
      value: webVitalsCallback,
    });
    Object.defineProperty(jsdom.window.webVitals, 'onFID', {
      value: webVitalsCallback,
    });
    Object.defineProperty(jsdom.window.webVitals, 'onINP', {
      value: webVitalsCallback,
    });
    Object.defineProperty(jsdom.window.webVitals, 'onTTFB', {
      value: webVitalsCallback,
    });

    const core = trackerCore({
      corePlugins: [],
      callback: (payloadBuilder) => {
        const [data, ...context] = payloadBuilder.getJson();
        expect(data.json.data).toMatchSnapshot();
        expect(context.map(({ json }) => json)).toMatchSnapshot('context');
        done();
      },
    });

    WebVitalsPlugin({
      loadWebVitalsScript: false,
      context: [{ schema: 'iglu:com.example/test/jsonschema/1-0-0', data: { ok: true } }],
    }).activateBrowserPlugin?.({ core } as BrowserTracker);
    const pagehideEvent = new PageTransitionEvent('pagehide');
    jsdom.window.dispatchEvent(pagehideEvent);
  });
});
