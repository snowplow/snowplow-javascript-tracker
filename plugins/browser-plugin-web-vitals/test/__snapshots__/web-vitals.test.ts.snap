// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Web Vitals plugin Returns values for Web Vitals properties 1`] = `
Object {
  "data": Object {
    "cls": 0.01,
    "fcp": 0.01,
    "fid": 0.01,
    "inp": 0.01,
    "lcp": 0.01,
    "navigationType": "navigation",
    "ttfb": 0.01,
  },
  "schema": "iglu:com.snowplowanalytics.snowplow/web_vitals/jsonschema/1-0-0",
}
`;

exports[`Web Vitals plugin Returns values for Web Vitals properties: context 1`] = `
Array [
  Object {
    "data": Array [
      Object {
        "data": Object {
          "ok": true,
        },
        "schema": "iglu:com.example/test/jsonschema/1-0-0",
      },
    ],
    "schema": "iglu:com.snowplowanalytics.snowplow/contexts/jsonschema/1-0-0",
  },
]
`;
